javascriptlib/src/mill/javascriptlib/PublishModule.scala

```diff
@@ -0,0 +1,539 @@
+  }
+
+  // Compilation Options
+
+  // EsBuild - Copying Resources
+  // we use tsc to compile to js & ts-patch to transform ts-paths
+  // esbuild script serves to copy resources to dist/
+  override def bundleScriptBuilder: Task[String] = Task.Anon {
+    def envName(input: String): String = {
+      val cleaned = input.replaceAll("[^a-zA-Z0-9]", "") // remove special characters
+      cleaned.toUpperCase
+    }
+
+    val copyPluginCode =
+      s"""
+         |  plugins: [
+         |    ${resources().map { p => p.path }.filter(os.exists).map { rp =>
+          s"""    copyStaticFiles({
+             |      src: ${ujson.Str(rp.toString)},
+             |      dest: ${ujson.Str(
+              publishDir().path.toString + "/" + bundledOut() + "/" + rp.last
+            )},
+             |      dereference: true,
+             |      preserveTimestamps: true,
+             |      recursive: true,
+             |    }),
+             """.stripMargin
+        }.mkString("\n")}
+         |    TsconfigPathsPlugin({ tsconfig: 'tsconfig.json' }),
+         |  ],
+         |""".stripMargin
+
+    val defineSection = resources().map { rp =>
+      val resourceRoot = rp.path.last
+      val envVarName = envName(resourceRoot)
+      s""" "process.env.$envVarName": JSON.stringify(${ujson.Str("./" + resourceRoot)})"""
+    }.distinct.mkString(",\n")
+
+    s"""|import * as esbuild from 'esbuild';
+        |import * as glob from 'glob';
+        |import TsconfigPathsPlugin from '@esbuild-plugins/tsconfig-paths'
+        |import copyStaticFiles from 'esbuild-copy-static-files';
+        |
+        |esbuild.build({
+        |  entryPoints: [],
+        |  outdir: ${ujson.Str("./" + bundledOut())},
+        |  write: ${ujson.Bool(false)}, // Prevent esbuild from generating new files
+        |  $copyPluginCode
+        |  define: {
+        |       $defineSection
+        |    },
+        |}).then(() => {
+        |  console.log('Build succeeded!');
+        |}).catch((e) => {
+        |  console.error('Build failed!');
+        |  console.error(e)
+        |  process.exit(1);
+        |});
+        |""".stripMargin
+
+  }
+
+  override def bundle: T[PathRef] = Task {
+    val tsnode = npmInstall().path / "node_modules/.bin/ts-node"
+    val bundleScript = compile()._1.path / "build.ts"
+    val bundle = Task.dest / "bundle.js"
+
+    os.write.over(
+      bundleScript,
+      bundleScriptBuilder()
+    )
+
+    os.call(
+      (tsnode, bundleScript),
+      stdout = os.Inherit,
+      cwd = compile()._1.path
+    )
+    PathRef(bundle)
+  }
+
+  // EsBuild - END
+
+  // publishDir; is used to process and compile files for publishing
+  def publishDir: T[PathRef] = Task { PathRef(Task.dest) }
+
+  def publish(): Command[CommandResult] = Task.Command {
+    // build package.json
+    packageJson()
+
+    // bundle code for publishing
+    bundle()
+
+    // run npm publish
+    os.call(("npm", "publish"), cwd = publishDir().path)
```

Comment:
Probably better to make this a `Command[Unit]` and make `os.call` have `stdout = os.Inherit` so users can see the logs